File: arodnap/doctor.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
import sys
from typing import Any, Literal

from arodnap.build_adapters import (
    AdapterExecutionError,
    MissingBuildToolError,
    ProjectModel,
    UnsupportedProjectError,
    select_build_adapter,
)
from arodnap.contracts import RunConfig
from arodnap.orchestrator.results import write_json
from arodnap.orchestrator.workspace import copied_workspace
from arodnap.patch_tool import PatchToolError, discover_patch_tool
from arodnap.analysis.checker_framework import (
    CheckerFrameworkError,
    resolve_analysis_jdk,
    tested_jdk_majors,
)


DoctorStatus = Literal["ok", "warning", "error"]


@dataclass(frozen=True)
class DoctorCheck:
    name: str
    status: DoctorStatus
    message: str
    details: dict[str, Any] | None = None

    def to_payload(self) -> dict[str, Any]:
        payload = {
            "name": self.name,
            "status": self.status,
            "message": self.message,
        }
        if self.details is not None:
            payload["details"] = _jsonable(self.details)
        return payload


@dataclass(frozen=True)
class DoctorReport:
    repo_root: Path
    out_dir: Path
    build_args: tuple[str, ...]
    compile_target: str | None
    success: bool
    checks: tuple[DoctorCheck, ...]

    def to_payload(self) -> dict[str, Any]:
        return {
            "repo_root": str(self.repo_root),
            "out_dir": str(self.out_dir),
            "build_args": list(self.build_args),
            "compile_target": self.compile_target,
            "success": self.success,
            "checks": [check.to_payload() for check in self.checks],
        }


def run_doctor(config: RunConfig) -> int:
    checks: list[DoctorCheck] = []
    checks.extend(_run_environment_checks(config))

    repo_path_check = _check_repo_path(config.repo_root)
    checks.append(repo_path_check)
    unsupported = _unsupported_project_check(config) if repo_path_check.status == "ok" else None
    if unsupported is not None:
        # No build Arodnap can capture: say so without copying the repository first.
        checks.append(unsupported)
    elif repo_path_check.status == "ok":
        try:
            with copied_workspace(config.repo_root, keep_workspace=config.keep_workspace) as workspace:
                checks.extend(
                    _run_repo_checks(
                        config,
                        workspace_root=workspace.workspace_root,
                    )
                )
        except OSError as exc:
            checks.append(
                DoctorCheck(
                    name="workspace_copy",
                    status="error",
                    message=f"Failed to create workspace copy for doctor: {exc}",
                    details={"repo_root": config.repo_root},
                )
            )

    success = all(check.status != "error" for check in checks)
    report = DoctorReport(
        repo_root=config.repo_root,
        out_dir=config.out_dir,
        build_args=tuple(config.build_args),
        compile_target=config.compile_target,
        success=success,
        checks=tuple(checks),
    )
    report_path = write_json(config.out_dir / "doctor.json", report.to_payload())
    _print_summary(report, report_path)
    return 0 if success else 1


def _run_environment_checks(config: RunConfig) -> list[DoctorCheck]:
    return [
        _check_python_runtime(),
        _check_java_runtime(config.cf_root),
        _check_patch_binary(),
        _check_checker_framework_path(config.cf_root),
        _check_checker_framework_tools(config.cf_root),
        _check_plugin_jars(config),
    ]


def _check_python_runtime() -> DoctorCheck:
    version = f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}"
    return DoctorCheck(
        name="python_runtime",
        status="ok",
        message=f"Python {version} is available.",
        details={"executable": Path(sys.executable), "version": version},
    )


def _check_java_runtime(cf_root: Path) -> DoctorCheck:
    try:
        jdk = resolve_analysis_jdk(cf_root)
    except CheckerFrameworkError as exc:
        return DoctorCheck(name="java_runtime", status="error", message=str(exc))

    details = {"home": jdk.home, "major_version": jdk.major_version, "source": jdk.source}
    tested = tested_jdk_majors(cf_root)
    if tested and jdk.major_version > max(tested):
        return DoctorCheck(
            name="java_runtime",
            status="warning",
            message=(
                f"JDK {jdk.major_version} at {jdk.home} is newer than the JDKs {cf_root.name} is tested on "
                f"(up to {max(tested)}). It may work; if analysis fails, use a tested JDK or a newer "
                "Checker Framework (--checker-framework)."
            ),
            details=details,
        )
    return DoctorCheck(
        name="java_runtime",
        status="ok",
        message=f"JDK {jdk.major_version} is available at {jdk.home} (from {jdk.source}).",
        details=details,
    )


def _check_patch_binary() -> DoctorCheck:
    try:
        tool = discover_patch_tool(require_gnu=False, operation_label="arodnap doctor")
    except PatchToolError as exc:
        return DoctorCheck(
            name="patch_binary",
            status="error",
            message=str(exc),
        )

    if tool.flavor == "gnu":
        status: DoctorStatus = "ok"
        message = f"GNU patch is available via {tool.binary}."
    else:
        status = "warning"
        message = (
            f"{tool.flavor.upper()} patch is available via {tool.binary}. "
            "GNU patch is preferred for repair-stage dry-run validation."
        )

    return DoctorCheck(
        name="patch_binary",
        status=status,
        message=message,
        details={"binary": tool.binary, "flavor": tool.flavor, "version": tool.version},
    )


def _check_checker_framework_path(cf_root: Path) -> DoctorCheck:
    if cf_root.is_dir():
        return DoctorCheck(
            name="checker_framework_path",
            status="ok",
            message=f"Vendored Checker Framework path is present: {cf_root}",
            details={"path": cf_root},
        )

    return DoctorCheck(
        name="checker_framework_path",
        status="error",
        message=f"Vendored Checker Framework path is missing: {cf_root}",
        details={"path": cf_root},
    )


def _check_checker_framework_tools(cf_root: Path) -> DoctorCheck:
    required = {
        "checker_jar": cf_root / "checker" / "dist" / "checker.jar",
    }
    missing = [name for name, path in required.items() if not path.is_file()]
    if not missing:
        return DoctorCheck(
            name="checker_framework_tools",
            status="ok",
            message="Checker Framework helper tools are present.",
            details=required,
        )

    return DoctorCheck(
        name="checker_framework_tools",
        status="error",
        message=f"Checker Framework helper tools are missing: {', '.join(sorted(missing))}",
        details=required,
    )


def _check_plugin_jars(config: RunConfig) -> DoctorCheck:
    required = {
        "close_injector_jar": config.close_injector_jar,
        "owning_field_jar": config.owning_field_jar,
        "rlfixer_jar": config.rlfixer_jar,
        "rlpatcher_jar": config.rlpatcher_jar,
    }
    missing = [name for name, path in required.items() if not path.is_file()]
    if not missing:
        return DoctorCheck(
            name="plugin_jars",
            status="ok",
            message="Required plugin jars are present.",
            details=required,
        )

    return DoctorCheck(
        name="plugin_jars",
        status="error",
        message=f"Required plugin jars are missing: {', '.join(sorted(missing))}",
        details=required,
    )


def _check_repo_path(repo_root: Path) -> DoctorCheck:
    if not repo_root.exists():
        return DoctorCheck(
            name="repo_path",
            status="error",
            message=f"Repository path does not exist: {repo_root}",
            details={"repo_root": repo_root},
        )
    if not repo_root.is_dir():
        return DoctorCheck(
            name="repo_path",
            status="error",
            message=f"Repository path is not a directory: {repo_root}",
            details={"repo_root": repo_root},
        )

    return DoctorCheck(
        name="repo_path",
        status="ok",
        message=f"Repository path exists: {repo_root}",
        details={"repo_root": repo_root},
    )


def _unsupported_project_check(config: RunConfig) -> DoctorCheck | None:
    """Adapter selection on the original repository; it only looks at which build files exist."""
    try:
        select_build_adapter(
            config.repo_root,
            compile_target=config.compile_target,
            build_args=config.build_args,
            build_command=config.build_command,
            timeouts=config.timeouts,
        )
    except UnsupportedProjectError as exc:
        return DoctorCheck(
            name="adapter_selection",
            status="error",
            message=str(exc),
            details={"repo_root": config.repo_root},
        )
    return None


def _run_repo_checks(
    config: RunConfig,
    *,
    workspace_root: Path,
) -> list[DoctorCheck]:
    checks: list[DoctorCheck] = []

    try:
        adapter = select_build_adapter(
            workspace_root,
            compile_target=config.compile_target,
            build_args=config.build_args,
            build_command=config.build_command,
            timeouts=config.timeouts,
        )
    except UnsupportedProjectError as exc:
        checks.append(
            DoctorCheck(
                name="adapter_selection",
                status="error",
                message=_normalize_workspace_message(str(exc), workspace_root=workspace_root, repo_root=config.repo_root),
                details={"repo_root": config.repo_root},
            )
        )
        return checks

    checks.append(
        DoctorCheck(
            name="adapter_selection",
            status="ok",
            message=f"Selected adapter {adapter.adapter_name} for build system {adapter.build_system}.",
            details={
                "adapter_name": adapter.adapter_name,
                "build_system": adapter.build_system,
            },
        )
    )

    try:
        project = adapter.inspect()
    except (MissingBuildToolError, UnsupportedProjectError, AdapterExecutionError) as exc:
        checks.append(
            DoctorCheck(
                name="repo_support",
                status="error",
                message=_normalize_workspace_message(str(exc), workspace_root=workspace_root, repo_root=config.repo_root),
                details={
                    "adapter_name": adapter.adapter_name,
                    "build_system": adapter.build_system,
                },
            )
        )
        return checks

    checks.append(
        DoctorCheck(
            name="repo_support",
            status="ok",
            message=_capture_summary(project),
            details=_project_details(project, workspace_root=workspace_root, repo_root=config.repo_root),
        )
    )
    checks.append(
        DoctorCheck(
            name="source_root",
            status="ok",
            message=f"Analysis root: {_display_repo_path(project.source_root, workspace_root=workspace_root, repo_root=config.repo_root)}",
            details={
                "source_root": _display_repo_path(project.source_root, workspace_root=workspace_root, repo_root=config.repo_root),
                "compiled_classes_root": _display_repo_path(
                    project.compiled_classes_root,
                    workspace_root=workspace_root,
                    repo_root=config.repo_root,
                ),
            },
        )
    )

    try:
        adapter.validate_compile(project)
    except (UnsupportedProjectError, AdapterExecutionError) as exc:
        checks.append(
            DoctorCheck(
                name="compile_target",
                status="error",
                message=_normalize_workspace_message(str(exc), workspace_root=workspace_root, repo_root=config.repo_root),
                details={
                    "compile_target": project.compile_target,
                    "build_tool": list(project.build_tool),
                    "build_tool_source": project.build_tool_source,
                },
            )
        )
        return checks

    checks.append(
        DoctorCheck(
            name="compile_target",
            status="ok",
            message=f"The project's {project.build_system} build compiled successfully ({project.build_tool_source} {project.build_tool[0]}).",
            details={
                "compile_target": project.compile_target,
                "build_tool": list(project.build_tool),
                "build_tool_source": project.build_tool_source,
                "compiled_classes_root": _display_repo_path(
                    project.compiled_classes_root,
                    workspace_root=workspace_root,
                    repo_root=config.repo_root,
                ),
            },
        )
    )
    return checks


def _capture_summary(project: ProjectModel) -> str:
    inputs = getattr(project, "inputs", None)
    if inputs is None:
        return f"Repository has a supported {project.build_system} build."
    generated = f" and {len(inputs.generated_sources)} generated" if inputs.generated_sources else ""
    return (
        f"Captured {len(inputs.units)} compile unit(s) with {len(inputs.sources)} source file(s){generated} "
        f"from the {project.build_system} build."
    )


def _project_details(
    project: ProjectModel,
    *,
    workspace_root: Path,
    repo_root: Path,
) -> dict[str, Any]:
    return {
        "adapter_name": project.adapter_name,
        "build_system": project.build_system,
        "build_file": _display_repo_path(project.build_file, workspace_root=workspace_root, repo_root=repo_root),
        "build_tool": list(project.build_tool),
        "build_tool_source": project.build_tool_source,
        "compile_target": project.compile_target,
        "source_root": _display_repo_path(project.source_root, workspace_root=workspace_root, repo_root=repo_root),
        "compiled_classes_root": _display_repo_path(
            project.compiled_classes_root,
            workspace_root=workspace_root,
            repo_root=repo_root,
        ),
    }


def _display_repo_path(path: Path, *, workspace_root: Path, repo_root: Path) -> Path:
    try:
        relative = path.resolve().relative_to(workspace_root.resolve())
    except ValueError:
        return path.resolve()
    return repo_root.resolve() / relative


def _normalize_workspace_message(message: str, *, workspace_root: Path, repo_root: Path) -> str:
    return message.replace(str(workspace_root.resolve()), str(repo_root.resolve()))


def _print_summary(report: DoctorReport, report_path: Path) -> None:
    for check in report.checks:
        print(f"[{check.status.upper()}] {check.name}: {check.message}")
    print(f"Doctor report: {report_path}")


def _jsonable(value: Any) -> Any:
    if isinstance(value, Path):
        return str(value)
    if isinstance(value, dict):
        return {key: _jsonable(item) for key, item in value.items()}
    if isinstance(value, (list, tuple)):
        return [_jsonable(item) for item in value]
    return value


__all__ = ["DoctorCheck", "DoctorReport", "run_doctor"]

File: arodnap-engine/src/main/java/org/arodnap/engine/doctor/Doctor.java
package org.arodnap.engine.doctor;

import java.io.IOException;
import java.io.PrintStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import org.arodnap.engine.files.FilePaths;
import org.arodnap.engine.inputs.ProjectCapture;
import org.arodnap.engine.inputs.ProjectCapture.CapturedProject;
import org.arodnap.engine.inputs.UnsupportedProjectException;
import org.arodnap.engine.jdk.Jdk;
import org.arodnap.engine.jdk.JdkLocator;
import org.arodnap.engine.json.Json;
import org.arodnap.engine.pipeline.RunSettings;
import org.arodnap.engine.pipeline.Workspace;
import org.arodnap.engine.tools.CheckerFramework;
import org.arodnap.engine.tools.Toolchain;
import org.arodnap.model.BuildDescription;

/**
 * Checks that Arodnap can run on a project: the JDK, the tools, and whether the project's inputs
 * can be captured (on a copy). Writes {@code doctor.json}; its check names are stable output.
 */
public final class Doctor {
    /** A check's outcome. */
    public enum Status { OK, WARNING, ERROR }

    /** One check. */
    public record Check(String name, Status status, String message, Map<String, Object> details) {
        Map<String, Object> toJson() {
            Map<String, Object> json = new LinkedHashMap<>();
            json.put("name", name);
            json.put("status", status.name().toLowerCase(Locale.ROOT));
            json.put("message", message);
            if (details != null) {
                json.put("details", details);
            }
            return json;
        }
    }

    private final Toolchain toolchain;
    private final JdkLocator jdks;

    public Doctor(Toolchain toolchain, JdkLocator jdks) {
        this.toolchain = toolchain;
        this.jdks = jdks;
    }

    /** Runs every check, writes doctor.json and prints one line per check; returns whether none failed. */
    public boolean run(RunSettings settings, ProjectCapture capture, PrintStream out) throws IOException {
        List<Check> checks = new ArrayList<>();
        checks.add(javaRuntime());
        checks.add(checkerFrameworkPath());
        checks.add(checkerFrameworkTools());
        checks.add(pluginJars());
        Check repoPath = repoPath(settings.repoRoot());
        checks.add(repoPath);
        if (repoPath.status() == Status.OK) {
            Check unsupported = null;
            try {
                capture.detect(settings.repoRoot());
            } catch (UnsupportedProjectException e) {
                // No build Arodnap can capture: say so without copying the repository first.
                unsupported = new Check("adapter_selection", Status.ERROR, e.getMessage(), details("repo_root", settings.repoRoot().toString()));
            }
            if (unsupported != null) {
                checks.add(unsupported);
            } else {
                try (Workspace workspace = Workspace.copyOf(settings.repoRoot(), settings.keepWorkspace())) {
                    checks.addAll(repoChecks(settings, capture, workspace));
                } catch (IOException e) {
                    checks.add(new Check("workspace_copy", Status.ERROR, "Failed to create workspace copy for doctor: " + e.getMessage(),
                            details("repo_root", settings.repoRoot().toString())));
                }
            }
        }
        boolean success = checks.stream().noneMatch(check -> check.status() == Status.ERROR);
        Map<String, Object> report = new LinkedHashMap<>();
        report.put("repo_root", settings.repoRoot().toString());
        report.put("out_dir", settings.outDir().toString());
        report.put("build_args", settings.buildArgs());
        report.put("compile_target", settings.compileTarget().orElse(null));
        report.put("success", success);
        report.put("checks", checks.stream().map(Check::toJson).toList());
        Path file = Json.writeFile(settings.outDir().resolve("doctor.json"), report, true);
        for (Check check : checks) {
            out.println("[" + check.status().name() + "] " + check.name() + ": " + check.message());
        }
        out.println("Doctor report: " + file);
        return success;
    }

    private Check javaRuntime() {
        Jdk jdk;
        try {
            jdk = CheckerFramework.analysisJdk(toolchain, jdks);
        } catch (CheckerFramework.CheckerFrameworkException e) {
            return new Check("java_runtime", Status.ERROR, e.getMessage(), null);
        }
        Map<String, Object> details = new LinkedHashMap<>();
        details.put("home", jdk.home().toString());
        details.put("major_version", jdk.majorVersion());
        details.put("source", jdk.source());
        if (toolchain.newestTestedJdk().isPresent() && jdk.majorVersion() > toolchain.newestTestedJdk().get()) {
            return new Check("java_runtime", Status.WARNING, "JDK " + jdk.majorVersion() + " at " + jdk.home()
                    + " is newer than the JDKs Checker Framework " + toolchain.checkerFrameworkVersion() + " is tested on (up to "
                    + toolchain.newestTestedJdk().get() + "). It may work; if analysis fails, use a tested JDK.", details);
        }
        return new Check("java_runtime", Status.OK, "JDK " + jdk.majorVersion() + " is available at " + jdk.home() + " (from " + jdk.source() + ").",
                details);
    }

    private Check checkerFrameworkPath() {
        Path directory = toolchain.checkerJar().toAbsolutePath().getParent();
        Map<String, Object> details = details("path", String.valueOf(directory));
        if (directory != null && Files.isDirectory(directory)) {
            return new Check("checker_framework_path", Status.OK, "Checker Framework " + toolchain.checkerFrameworkVersion() + " is present: "
                    + directory, details);
        }
        return new Check("checker_framework_path", Status.ERROR, "Checker Framework is missing: " + directory, details);
    }

    private Check checkerFrameworkTools() {
        Map<String, Object> details = details("checker_jar", toolchain.checkerJar().toString());
        if (Files.isRegularFile(toolchain.checkerJar())) {
            return new Check("checker_framework_tools", Status.OK, "Checker Framework tools are present.", details);
        }
        return new Check("checker_framework_tools", Status.ERROR, "Checker Framework tools are missing: checker_jar", details);
    }

    private Check pluginJars() {
        Map<String, Object> details = new LinkedHashMap<>();
        toolchain.files().forEach((name, path) -> details.put(name, String.valueOf(path)));
        details.put("stubs_directory", toolchain.stubsDirectory().toString());
        List<String> missing = toolchain.missing().stream().filter(name -> !name.equals("checker_jar")).toList();
        if (missing.isEmpty()) {
            return new Check("plugin_jars", Status.OK, "Required tool jars are present.", details);
        }
        return new Check("plugin_jars", Status.ERROR, "Required tool jars are missing: " + String.join(", ", missing), details);
    }

    private static Check repoPath(Path repoRoot) {
        Map<String, Object> details = details("repo_root", repoRoot.toString());
        if (!Files.exists(repoRoot)) {
            return new Check("repo_path", Status.ERROR, "Repository path does not exist: " + repoRoot, details);
        }
        if (!Files.isDirectory(repoRoot)) {
            return new Check("repo_path", Status.ERROR, "Repository path is not a directory: " + repoRoot, details);
        }
        return new Check("repo_path", Status.OK, "Repository path exists: " + repoRoot, details);
    }

    private static List<Check> repoChecks(RunSettings settings, ProjectCapture capture, Workspace workspace) {
        List<Check> checks = new ArrayList<>();
        Path repoRoot = settings.repoRoot();
        BuildDescription build;
        try {
            build = capture.detect(workspace.root());
        } catch (UnsupportedProjectException e) {
            checks.add(new Check("adapter_selection", Status.ERROR, inRepo(e.getMessage(), workspace), details("repo_root", repoRoot.toString())));
            return checks;
        }
        Map<String, Object> selection = new LinkedHashMap<>();
        selection.put("adapter_name", build.adapterName());
        selection.put("build_system", build.buildSystem());
        checks.add(new Check("adapter_selection", Status.OK,
                "Selected adapter " + build.adapterName() + " for build system " + build.buildSystem() + ".", selection));

        CapturedProject captured;
        try {
            captured = capture.capture(workspace.root(), workspace.stateDirectory());
        } catch (UnsupportedProjectException e) {
            checks.add(new Check("repo_support", Status.ERROR, inRepo(e.getMessage(), workspace), selection));
            return checks;
        }
        BuildDescription captureBuild = captured.build();
        int generated = captured.inputs().generatedSources().size();
        Map<String, Object> project = new LinkedHashMap<>();
        project.put("adapter_name", captureBuild.adapterName());
        project.put("build_system", captureBuild.buildSystem());
        project.put("build_file", display(captureBuild.buildFile(), workspace));
        project.put("build_tool", captureBuild.buildTool());
        project.put("build_tool_source", captureBuild.buildToolSource());
        project.put("compile_target", captureBuild.compileTarget());
        project.put("source_root", display(captured.inputs().sourceRoot(), workspace));
        checks.add(new Check("repo_support", Status.OK, "Captured " + captured.inputs().units().size() + " compile unit(s) with "
                + captured.inputs().sources().size() + " source file(s)" + (generated > 0 ? " and " + generated + " generated" : "")
                + " from the " + captureBuild.buildSystem() + " build.", project));
        checks.add(new Check("source_root", Status.OK, "Analysis root: " + display(captured.inputs().sourceRoot(), workspace),
                details("source_root", display(captured.inputs().sourceRoot(), workspace))));
        String tool = captureBuild.buildTool().isEmpty() ? captureBuild.adapterName() : captureBuild.buildTool().get(0);
        Map<String, Object> compile = new LinkedHashMap<>();
        compile.put("compile_target", captureBuild.compileTarget());
        compile.put("build_tool", captureBuild.buildTool());
        compile.put("build_tool_source", captureBuild.buildToolSource());
        checks.add(new Check("compile_target", Status.OK, "The project's " + captureBuild.buildSystem() + " build compiled successfully ("
                + captureBuild.buildToolSource() + " " + tool + ").", compile));
        return checks;
    }

    /** A path in the copy, shown as the same path in the project. */
    private static String display(Path path, Workspace workspace) {
        return FilePaths.relativeTo(path, workspace.root()).map(relative -> workspace.repoRoot().resolve(relative).toString())
                .orElse(FilePaths.real(path).toString());
    }

    private static String inRepo(String message, Workspace workspace) {
        return message == null ? "" : message.replace(workspace.root().toString(), workspace.repoRoot().toString());
    }

    private static Map<String, Object> details(String key, Object value) {
        Map<String, Object> details = new LinkedHashMap<>();
        details.put(key, value);
        return details;
    }
}
